cylc/flow/xtrigger_mgr.py

```diff
@@ -489,23 +511,9 @@ def callback(self, ctx: SubFuncContext):
             return
         LOG.debug('%s: returned %s', sig, results)
         if satisfied:
+            # Newly satisfied
             self.data_store_mgr.delta_task_xtrigger(sig, True)
+            self.workflow_db_mgr.put_xtriggers({sig: results})
```

Comment:
I'm a bit confused by what should and shouldn't be in the xtriggers table.

The `put_xtriggers` routine wipes the entire table, then adds the new entry. If you call `put_xtriggers` multiple times *before* the workflow DB queue is processed, all of the wiping will happen first, and all the adding second.

https://github.com/cylc/cylc-flow/blob/df350cd5481f8e42f3e236840816ca296d855660/cylc/flow/workflow_db_mgr.py#L406-L412

Practical upshot, anything added to the xtrigger table will get deleted the next time an xtrigger is satisfied in a future main loop cycle?

Reply:
> I'm a bit confused by what should and shouldn't be in the xtriggers table.

It should contain the function signature and result of all satisfied xtriggers - i.e. ones for which the function no longer needs to be called, because we already have the result.

Reply:
>  it looks like there are duplicate SQL calls updating the broadcast tables (before and after). Guessing this ain't meant to happen?

They're not duplicate, each is for a different cycle point.

If multiple tasks depend on the same xtrigger, we broadcast its result to all of the dependent tasks.

So I don't think there's a problem.

Reply:
> This looks like a bug to me, e.g. an xtrigger could get re-run on restart, potentially causing different data to be broadcasted into the cycle?

Yikes, good spot!  That's a different bug, although possibly this branch exposed it more.  It seems to me the table delete update does not need to be there. I've removed it, and the table content and sqlite trace looks right now, for a multi-cycle run with variously-specialized xtriggers.

I guess we need a test to ensure that all satisfied xtriggers get stored ...

Reply:
> They're not duplicate, each is for a different cycle point.

Derp :facepalm: 

> Yikes, good spot! That's a different bug, although possibly this branch exposed it more

Yep, that was my conclusion.

> the table delete update does not need to be there

Without the delete, this table will accumulate indefinitely, which may be ok, so long as it doesn't mess with reflow use cases.

Presumably we could remove satisfied xtriggers from the table once the task has left the pool?

Reply:
> I'm guessing this used to work by dumping all xtriggers for all tasks in the pool into the DB for each main loop cycle. So the delete was effectively just removing tasks no longer in the pool?

Yes, I think so. BUT, given that external triggers are "external", their result should in principle be independent of the workflow state, so it seems to me they shouldn't need to be re-run if the dependent tasks get re-triggered after dropping out of the task pool. What do you think?

Reply:
That's fine by me, it will inflate the DB a bit, but it might actually speed up the DB operations (because we're no longer wiping the table every main loop cycle). So long as we store the flow numbers with the xtrigger it should be fine.

However, retries (which are implemented via xtriggers) might be a separate case?

Reply:
> However, retries (which are implemented via xtriggers) might be a separate case?

Retries are implemented by clock xtriggers, so they have wallclock time-specific function signatures, and so an old task that gets retriggered will still have correct retry delays.

```console
~/c/xtr $ sqlite3 ~/cylc-run/xtr/runN/.service/db 'select * from xtriggers'
wall_clock(trigger_time=1704799660.2086887)|{}  # original job (fail then succeed)
wall_clock(trigger_time=1704799695.2978518)|{}  # retriggered later (fail then succeed).
```

Hmmm. Which does suggest, however: https://github.com/cylc/cylc-flow/issues/5911